### PR TITLE
Added 'deprecated' field

### PR DESCRIPTION
**Pull request by mickeyn**
Added to both release and file types (which is also /module in API)
Field will be set by the 'release' script.

This PR is to address #572 

Note: after merge - need to run a one-off run to update field for existing releases

### CODE REVIEW

File: lib/MetaCPAN/Document/File.pm [hunk 1]
```diff
@@ -32,6 +32,20 @@ sub BUILD {
...
+
+Indicates file deprecation (this will normally mean the abstract will contain
+DEPRECATED / DEPRECIATED or the release is marked with x_deprecated = 1
+
+=cut
+
+has deprecated => (
+    is      => 'ro',
+    isa     => Bool,
+    default => sub {0},
```
Comment (oalders) [inline lib/MetaCPAN/Document/File.pm]: You should also be able to set a `Bool` default via `default => 0`,

File: lib/MetaCPAN/Document/Release.pm [hunk 2]
```diff
@@ -104,6 +104,12 @@ This is an ArrayRef of modules that are included in this release.
 
 =cut
 
+=head2 deprecated
+
+This is a boolean indicated whether the release is marked as deprecated
```
Comment (oalders) [inline lib/MetaCPAN/Document/Release.pm]: s/indicated/indicating/

File: lib/MetaCPAN/Script/Release.pm [hunk 3]
```diff
@@ -254,6 +263,12 @@ sub import_archive {
         $file->clear_module if ( $file->is_pod_file );
         $file->documentation;
         $file->suggest;
+
+        # check for DEPRECATED/DEPRECIATED in abstract of file
+        $file->_set_deprecated(1)
+            if $deprecated
+            or $file->abstract and $file->abstract =~ /DEPRECI?ATED/;
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: +1 for searching for the misspelling.

File: lib/MetaCPAN/Script/Release.pm [hunk 4]
```diff
@@ -237,6 +237,15 @@ sub import_archive {
...
     }
 
+# check for release deprecation in abstract of release or has x_deprecated in meta
+    my $deprecated = (
+               $meta->{x_deprecated}
+            or $document->has_abstract
+            and $document->abstract =~ /DEPRECI?ATED/
+    ) ? 1 : 0;
+
+    $document->_set_deprecated($deprecated);
```
Comment (oalders) [inline lib/MetaCPAN/Script/Release.pm]: First off, thanks so much for doing this.  I have wanted to see something like this for a long time.  My only real issue with the logic is that it seems to blur the lines between a release which is deprecated and a release which contains only *some* deprecated modules.  See https://metacpan.org/pod/release/ETHER/Dist-Zilla-Plugin-Deprecated-0.007/lib/Dist/Zilla/Plugin/Deprecated.pm#DESCRIPTION

Basically, checking for presence of the key will not always be quite enough.  If a release has 10 modules and one of them has been deprecated using `x_deprecated` key, then it looks like the entire release will be flagged.

File: lib/MetaCPAN/Script/Release.pm (hunk 3 above)
Comment (oalders) [reply lib/MetaCPAN/Script/Release.pm]: Based on the above comments, I think we need a bit more logic around flagging files as having been deprecated.  Are there any corner cases here we should note, like files containing inner packages where some could theoretically have been deprecated but others not?  Not sure how likely this is, but if it's possible, then it's worth noting.

File: lib/MetaCPAN/Script/Release.pm (hunk 4 above)
Comment (mickeyn) [reply lib/MetaCPAN/Script/Release.pm]: OK, that clarifies the issue I ran into with this :)
I'll try to fix the logic.

Comment (mickeyn) [reply lib/MetaCPAN/Script/Release.pm]: This one is indicating a release deprecation, i.e. the x_deprecated is in the top level of meta.
A single deprecated file does not set this flag, in that case we will have the x_deprecated key set under `meta->provides->file->x_deprecated` so that logic is just missing for the file later on - will add that.

Comment (mickeyn) [reply lib/MetaCPAN/Script/Release.pm]: added the mentioned check on file level

File: lib/MetaCPAN/Script/Release.pm (hunk 3 above)
Comment (mickeyn) [reply lib/MetaCPAN/Script/Release.pm]: ok, so now the file's check for x_deprecated is in place... this is still a check for any of the file's modules being deprecated as we are marking the file and not a single module (I guess that's a fair compromise in this case).

So other files will not be affected (check using: `bin/metacpan release --latest https://cpan.metacpan.org/authors/id/E/ET/ETHER/App-Nopaste-1.011.tar.gz` - see only `Codepeek.pm` gets deprecated) 

The open question I have is - do we want to mark all the file entries as deprecated when the release is marked as deprecated?

Comment (oalders) [reply lib/MetaCPAN/Script/Release.pm]: > The open question I have is - do we want to mark all the file entries as deprecated when the release is marked as deprecated?

I think so.

**Comment by oalders** [comment]
@karenetheridge please feel free to review/comment on this.

File: lib/MetaCPAN/Document/File.pm [hunk 5]
```diff
@@ -32,6 +32,20 @@ sub BUILD {
 
 =head1 PROPERTIES
 
+=head2 deprecated
+
+Indicates file deprecation (this will normally mean the abstract will contain
```
Comment (karenetheridge) [inline lib/MetaCPAN/Document/File.pm]: Will the file get the 'deprecated' flag if the entire release is deprecated, as well as the individual file is deprecated? If so, the documentation isn't quite right: it would be something more like:

    Indicates file deprecation (the abstract contains "DEPRECATED" or "DEPRECIATED", or the x_deprecated flag is included in the corresponding "provides" entry in distribution metadata); it is also set if the entire release is marked deprecated (see L<MetaCPAN::Document::Release#deprecated>).

File: lib/MetaCPAN/Document/Release.pm [hunk 6]
```diff
@@ -104,6 +104,12 @@ This is an ArrayRef of modules that are included in this release.
 
 =cut
 
+=head2 deprecated
+
+This is a boolean indicating whether the release is marked as deprecated
+
```
Comment (karenetheridge) [inline lib/MetaCPAN/Document/Release.pm]: I would add:

   ... (the main module's ABSTRACT contains "DEPRECATED" or "DEPRECIATED", or the x_deprecated flag is set in metadata)

File: lib/MetaCPAN/Document/File.pm (hunk 5 above)
Comment (mickeyn) [reply lib/MetaCPAN/Document/File.pm]: added as suggested

File: lib/MetaCPAN/Document/Release.pm (hunk 6 above)
Comment (mickeyn) [reply lib/MetaCPAN/Document/Release.pm]: added as suggested

**Comment by mickeyn** [comment]
thanks @karenetheridge

**Comment by mickeyn** [comment]
updated the mapping in production - now it's safe to merge
